compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -260,98 +251,43 @@ impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
 
     fn new(
         tcx: TyCtxt<'tcx>,
-        body: &'a mir::Body<'tcx>,
+        (body, body_id): (&'a thir::Thir<'tcx>, thir::ExprId),
     ) -> Result<Option<AbstractConstBuilder<'a, 'tcx>>, ErrorReported> {
-        let mut builder = AbstractConstBuilder {
-            tcx,
-            body,
-            nodes: IndexVec::new(),
-            locals: IndexVec::from_elem(NodeId::MAX, &body.local_decls),
-            checked_op_locals: BitSet::new_empty(body.local_decls.len()),
-        };
-
-        // We don't have to look at concrete constants, as we
-        // can just evaluate them.
-        if !body.is_polymorphic {
-            return Ok(None);
-        }
+        let builder =
+            AbstractConstBuilder { tcx, body_id, body: Lrc::new(body), nodes: IndexVec::new() };
 
-        // We only allow consts without control flow, so
-        // we check for cycles here which simplifies the
-        // rest of this implementation.
-        if body.is_cfg_cyclic() {
-            builder.error(None, "cyclic anonymous constants are forbidden")?;
+        struct IsThirPolymorphic<'a, 'tcx> {
+            is_poly: bool,
+            thir: &'a thir::Thir<'tcx>,
+            tcx: TyCtxt<'tcx>,
         }
 
-        Ok(Some(builder))
-    }
-
-    fn add_node(&mut self, node: Node<'tcx>, span: Span) -> NodeId {
-        // Mark used nodes.
-        match node {
-            Node::Leaf(_) => (),
-            Node::Binop(_, lhs, rhs) => {
-                self.nodes[lhs].used = true;
-                self.nodes[rhs].used = true;
-            }
-            Node::UnaryOp(_, input) => {
-                self.nodes[input].used = true;
-            }
-            Node::FunctionCall(func, nodes) => {
-                self.nodes[func].used = true;
-                nodes.iter().for_each(|&n| self.nodes[n].used = true);
+        use thir::visit;
+        impl<'a, 'tcx: 'a> visit::Visitor<'a, 'tcx> for IsThirPolymorphic<'a, 'tcx> {
+            fn thir(&self) -> &'a thir::Thir<'tcx> {
+                &self.thir
             }
-            Node::Cast(_, operand, _) => {
-                self.nodes[operand].used = true;
-            }
-        }
 
-        // Nodes start as unused.
-        self.nodes.push(WorkNode { node, span, used: false })
-    }
+            fn visit_expr(&mut self, expr: &thir::Expr<'tcx>) {
+                self.is_poly |= expr.ty.definitely_has_param_types_or_consts(self.tcx);
```

Comment:
`ExprKind::Cast` also contains a `Ty` and afaict you don't look at `thir::Pat` at all :thinking: 

maybe it makes sense to add a `fn visit_ty` to the `Visitor` trait and only manually implement `visit_ty` and `visit_const` here.

Reply:
ExprKind::Cast just contains an `ExprId` the type of the cast is from `expr.ty` so that's correct.

Reply:
I did forget about `thir::Pat` i think I thought it was unnecessary as i dont think you can currently use a pattern in a generic constant but that doesn't matter for checking this

Reply:
I'm not sure that visiting patterns is actually important, I cant think of anything in patterns that could be generic that wouldnt be handled by either `visit_const` from when a pattern is a constant or `visit_expr` for what's being `match`ed on/`let =` on

Reply:
woups, meant `ExprKind::Call` :sweat_smile: 

I personally don't think `thir::Pat` is too important either, or well, only for `is_polymorphic` (which we should be able to remove from `mir::Body` now). But the current impls seems fragile if we ever add a `Ty` to some other `ExprKind` or something, so i would like a safer approach here

Reply:
oh right, I assumed that the ty in `ExprKind::Call` was the same as the ty for the `fun` expr which is why I didnt visit that

Reply:
I changed how we check if thir is polymorphic to explicitly match on ExprKind and PatKind some of the arms are doing `definitely_has_param_types_or_consts` but im not really sure it actually makes any difference to be doing so

Reply:
> two match statements that exhaustively match on ExprKind and PatKind without actually doing anything which seems incredibly silly

:thinking: that doesn't make me too happy. not sure how easy and useful it is to add tests for this, but I think that having an exhaustive match here is more annoying than helpful '^^

imo its fine to just remove the match, we have enough time on nightly to see whether this is a problem